First release (#1)
AMRC-FactoryPlus/documentation: docs/framework-components/core-components/edge-agent.mdx

Reviewer: I've added a proposed diagram below

            <div className="p-4 bg-brand-20 dark:bg-brand-90 dark:text-white text-brand flex items-center justify-center opacity-40 ml-2">Directory</div>
            <div className="p-4 bg-brand-20 dark:bg-brand-90 dark:text-white text-brand flex items-center justify-center opacity-40">Configuration Store</div>
            <div className="p-4 bg-brand-20 dark:bg-brand-90 dark:text-white text-brand flex items-center justify-center opacity-40">Manager</div>
            <div className="p-4 bg-brand-20 dark:bg-brand-90 dark:text-white text-brand flex items-center justify-center opacity-40">Commands</div>
            <div className="p-4 bg-brand-20 dark:bg-brand-90 dark:text-white text-brand flex items-center justify-center opacity-40">Data Warehouse</div>
            <div className="p-4 bg-brand-20 dark:bg-brand-90 dark:text-white text-brand flex items-center justify-center opacity-40 col-span-full ml-2">MQTT</div>
            <div className="p-4 bg-brand-80 dark:bg-brand-40 text-white dark:text-brand flex items-center justify-center col-span-full ml-2">Edge Agents</div>
        </div>
    </div>
);

> The Edge Agent is the workhorse of Factory+, operating at the periphery, gathering data from devices, normalising it into Schemas, converting the packets into Sparkplug messages, and transmitting them to the MQTT component.

Reviewer: ```suggestion
The Edge Agent is the workhorse of Factory+, operating at the periphery, gathering data from devices, normalising it into [Data Schemas](/docs/schemas), converting the packets into Sparkplug messages, and transmitting them to the MQTT component.
```

> The Edge Agent is the workhorse of Factory+, operating at the periphery, gathering data from devices, normalising it into Schemas, converting the packets into Sparkplug messages, and transmitting them to the MQTT component.

Reviewer: ```suggestion
The Edge Agent is the workhorse of Factory+, its duties involve operating at the periphery, gathering data from devices, normalising it into Schemas, converting the packets into Sparkplug messages, and transmitting them to the MQTT component.
```

<OpenSourceExample buttonText={'ACS Edge Agent Component'} repoUrl={'https://github.com/AMRC-FactoryPlus/acs-edge'}></OpenSourceExample>

## Overview

<ComponentOverview></ComponentOverview>

Although the Edge Agent plays a crucial role in the framework there are few moving parts because all configuration is authored in the [Manager Component](/docs/framework-components/core-components/manager); the Edge Agent simply follows orders.

> 

Reviewer: ```suggestion
  
 ```mermaid
graph LR
  subgraph "Edge"
    A[Device] -- "OPC UA (Poll/Response)" --> B[Cell Gateway]
  end
  B -. "MQTT (Report by Exception)" .-> C[MQTT]
  D[Manager] -- "Configuration" -> A
```
  
```

Implementations of Edge Agents:
- **MUST** conform to the [Sparkplug V3 specification](https://sparkplug.eclipse.org/specification/version/3.0/) (i.e. they **MUST** utilise the Sparkplug payload model **MUST** publish to the correct Sparkplug topics, they **MUST** encode packets using Protocol Buffers and they **MUST** register/publish `DDEATH`/`NDEATH` and `DBIRTH`/`NBIRTH` messages).
- **SHOULD** be ephemeral, immutable and idempotent

> - **MUST** immediately request their configuration from the [Manager Component](/docs/framework-components/core-components/manager) using the `/api/edge-agent-config` endpoint when it starts. The configuration returned will have been validated by the manager to ensure that it conforms to the [Edge Agent Configuration Schema](https://github.com/AMRC-FactoryPlus/schemas/blob/main/Edge_Agent_Config.json), but Edge Agents **SHOULD** also validate that this is the case before they load the configuration.

Reviewer: ```suggestion
- **MUST** immediately request their configuration from the [Manager Component](/docs/framework-components/core-components/manager) using the `/api/edge-agent-config` endpoint when it starts. The configuration returned will have been validated by the Manager to ensure that it conforms to the [Edge Agent Configuration Schema](https://github.com/AMRC-FactoryPlus/schemas/blob/main/Edge_Agent_Config.json), but Edge Agents **SHOULD** also validate that this is the case before they load the configuration.
```

Before the configuration is loaded, all sensitive information placeholders **MUST** be replaced with the real values, which **MUST** be made available to the Edge Agent at runtime.

:::info Secret Management
The [AMRC Connectivity Stack](https://github.com/AMRC-FactoryPlus/amrc-connectivity-stack) uses Kubernetes Secrets to manage sensitive information and make it available at the point of consumption.
:::

Once the configuration has been processed, the Edge Agent:
- **MUST** use the credentials in the configuration to connect to the MQTT component
- **MUST** publish `NBIRTH` and `DBIRTH` messages as appropriate
- **MUST** publish metrics on a report-by-exception basis, that is, only when a value has changed

> - **MUST** publish metric names as a slash (`/`) delimited string of the nested metric name (e.g. `/Phases/1/Voltage`), which **MUST** conform to the schema for the specific device

Reviewer: ```suggestion
- **MUST** publish metric names as a slash (`/`) delimited string of the nested metric name (e.g. `/Phases/1/Voltage`), which **MUST** conform to the [schema](/docs/schemas) for the specific device
```

- **MUST NOT** publish metrics that do not conform to a device schema

> - **MUST** respond to `Node Control/Reload Edge Agent Config` NCMDs messages, which **MUST** instruct the Edge Agent to fetch a fresh configuration from the Manager

Reviewer: ```suggestion
- **MUST** respond to `Node Control/Reload Edge Agent Config` NCMDs messages, which **MUST** instruct the Edge Agent to fetch the latest configuration from the Manager
```
